File: apps/sandbox-ts/process/handle.ts
import { ChildProcess } from "child_process"
import { Broadcaster } from "./broadcaster"
import { Writable } from "stream"

export interface ProcessHandle{
    id: string,
    process: ChildProcess,
    state: "running" | "exited" | "failed",
    stdin: Writable
    stdout: Broadcaster<Buffer>,
    stderr: Broadcaster<Buffer>,
    exitCode? : number,
    error? : string
}

File: apps/sandbox-go/internal/process/handle.go
package process

import (
	"io"
	"os/exec"
	"sandbox-go/internal/filesystem"
	"sync"
	"time"
)

type Handle struct {
	ID  string
	Cmd *exec.Cmd

	CreatedAt time.Time
	LastIOAt  time.Time
	TTL       time.Duration

	Stdin  io.WriteCloser
	Stdout *Broadcaster
	Stderr *Broadcaster

	FS *filesystem.Filesystem

	ExitCode int
	State    State
	Err      error

	mu sync.Mutex
}
